src/main/java/com/puppycrawl/tools/checkstyle/SarifLogger.java

```diff
@@ -156,6 +176,58 @@ public SarifLogger(
             readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultFileOnly.template");
         resultErrorOnly =
             readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultErrorOnly.template");
+        rule = readResource("/com/puppycrawl/tools/checkstyle/sarif/Rule.template");
+    }
+
+    /**
+     * Loads all available module metadata from XML files.
+     */
+    private void loadModuleMetadata() {
+        final List<ModuleDetails> allModules =
+                XmlMetaReader.readAllModulesIncludingThirdPartyIfAny();
+        allModuleMetadata = new LinkedHashMap<>(allModules.size());
+        for (ModuleDetails module : allModules) {
+            allModuleMetadata.put(module.getFullQualifiedName(), module);
+        }
+    }
+
+    /**
+     * Finds and caches module metadata for a given rule ID.
+     *
+     * @param ruleId the rule ID (violation key)
+     */
+    private void cacheRuleMetadata(String ruleId) {
+        if (!ruleMetadata.containsKey(ruleId)) {
+            for (ModuleDetails module : allModuleMetadata.values()) {
+                if (module.getViolationMessageKeys().contains(ruleId)) {
+                    ruleMetadata.put(ruleId, module);
+                    break;
+                }
+            }
+        }
+    }
+
+    /**
+     * Generates the JSON representation of rules for the SARIF report.
+     *
+     * @return JSON string representing the rules array
+     */
+    private String generateRulesJson() {
+        String rulesJson = "";
+        if (!ruleMetadata.isEmpty()) {
+            final List<String> ruleJsonList = new ArrayList<>(ruleMetadata.size());
+            for (Map.Entry<String, ModuleDetails> entry : ruleMetadata.entrySet()) {
+                final String ruleId = entry.getKey();
+                final ModuleDetails module = entry.getValue();
+                final String ruleJson = rule
+                        .replace(RULE_ID_PLACEHOLDER, escape(ruleId))
+                        .replace("${shortDescription}", escape(module.getName()))
+                        .replace("${fullDescription}", escape(module.getDescription()));
+                ruleJsonList.add(ruleJson);
+            }
+            rulesJson = "\n" + String.join(COMMA_LINE_SEPARATOR, ruleJsonList);
```

Comment:
```suggestion
            rulesJson = "\n" + String.join(COMMA_LINE_SEPARATOR, ruleMetadata.entrySet().map(this::toFoo));
```
use lambda giving `map` upfront, thats whats actually going on.